### PR TITLE
first pass at wiring up jito-plugin

### PR DESCRIPTION
**Pull request by segfaultdoc**
Allows code to be plugged in with access to consensus objects. Plugin at your own risk!

### CODE REVIEW

File: jito-plugin/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+[package]
+name = "jito-plugin"
```
Comment (buffalu) [inline jito-plugin/Cargo.toml:2]: what do you think about other names so we can upstream to solana labs?

for instance:
- runtime plugin
- execution plugin
- consensus plugin
- banking plugin

File: jito-plugin/src/jito_plugin.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    #[error("Failed to load a jito plugin")]
+    FailedToLoadPlugin(#[from] Box<dyn std::error::Error>),
+}
+
+// TODO: Probably need to repr(C) all nested structs
+#[repr(C)]
+pub struct PluginDependencies {
+    pub bank_forks: Arc<RwLock<BankForks>>,
+    pub block_commitment_cache: Arc<RwLock<BlockCommitmentCache>>,
```
[the remark this answers is not in the document]
Comment (buffalu) [reply jito-plugin/src/jito_plugin.rs:35]: maybe blockstore as well?

File: jito-plugin/Cargo.toml (hunk 1 above)
Comment (segfaultdoc) [reply jito-plugin/Cargo.toml:2]: runtime-plugin sgtm, doesn't look like labs wanna upstream

File: jito-plugin/src/jito_plugin.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (segfaultdoc) [reply jito-plugin/src/jito_plugin.rs:35]: maybe later, can get slot from working bank rn ++ no use for it atm. wanna minimize dependencies until absolutely needed

File: validator/src/main.rs [hunk 3]
```diff
@@ -1979,6 +2006,7 @@ pub fn main() {
         cluster_entrypoints,
         &validator_config,
         should_check_duplicate_instance,
+        Some(runtime_plugin_config_and_rpc_rx),
```
Comment (buffalu) [inline validator/src/main.rs:2009]: nit: add to end of args

File: runtime-plugin/src/runtime_plugin_service.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            Self::start_rpc_request_handler(rpc_receiver, plugin_manager.clone(), exit.clone());
+
+        Ok(Self {
+            plugin_manager,
+            rpc_thread,
+        })
+    }
+
+    pub fn join(self) {
+        self.rpc_thread.join().unwrap();
```
Comment (buffalu) [inline runtime-plugin/src/runtime_plugin_service.rs:55]: propagate error so shutdown can happen properly

File: core/src/validator.rs [hunk 5]
```diff
@@ -889,6 +897,17 @@ impl Validator {
...
 
+        if let Some((runtime_plugin_configs, request_rx)) = runtime_plugin_configs_and_request_rx {
+            RuntimePluginService::start(
+                &runtime_plugin_configs,
+                request_rx,
+                bank_forks.clone(),
+                block_commitment_cache.clone(),
+                exit.clone(),
+            )
+            .unwrap();
```
Comment (buffalu) [inline core/src/validator.rs:908]: map err

**Review by buffalu — APPROVED** [review]

**Pull request merged**